### DISCUSSION TITLE
Composite keybindings + scripting from shell

### DISCUSSION THREAD

**Discussion by s-valent** [ghostty-org/ghostty#2872] [Feature Requests, Ideas]
It would be awesome if ghostty supported more complex keybind actions, maybe such as alacritty's `command = { program ... }`. 

Here is an example of alacritty's shortcut to switch to previous job (I haven't found a way to do it with alacritty's built in keymaps sequences since it requires a delay):
```toml
[[keyboard.bindings]]
key = "["
mods = "Command"
command = { program = "/usr/bin/osascript", args = ['-e', 'tell application "System Events" to keystroke "z" using control down', '-e', 'delay 0.1', '-e', 'tell application "System Events" to keystroke "[" using control down'] }
```

```bash
bind -x '"\C-[": fg %-'
```

With the additional ability to call actions from ghostty cli (for example `ghostty -a write_scrollback_file:paste`, similar to how it is done in tmux) it could enable some potentially interesting scenarios, such as parsing scrollback and opening each file in vim in new tab (or just the file under the cursor).

Potential examples:

```bash
# ~/.config/ghostty/config
keybind = "cmd+[=command:switch_to_previous_job.sh"
keybind = "cmd+o=command:open_file_in_new_tab.sh"
keybind = "cmd+f=command:search.sh"

# switch_to_previous_job.sh
ghostty -a 'text:\x1A'
sleep 0.3
ghostty -a 'text:fg %-\r'

# open_file_in_new_tab.sh
ghostty -a write_scrollback_file | head -n $(ghostty -a get_cursor_position | cut -f 2 -d:) | tail -n1 | xargs ghostty -e vim

# search.sh
file=$(ghostty -a write_scrollback_file)
ghostty -e new_window
ghostty -e resize_window:500,500
ghostty -e toggle_window_decorations
ghostty -e "text:vim $file\r"
```

**Comment by neurocyte** [comment]
I rely heavily on keybindings that launch commands in a new tab/window based off the current window's cwd. This is really core to my daily workflow.

Here's a common example that I use in kitty:
```
map kitty_mod+a>kitty_mod+d launch --cwd=last_reported --type=tab lazydocker
```

The lack of launch keybindings is the only reason I can't really use ghostty as my main daily driver terminal.

**Comment by user529** [comment]
And one more vote for keybind-scripting-feature.

Another option that could fit me would be a "custom profile" feature. It would works like keybinding that opens a new tab + startup command or startup script.

**Comment by SuperDuperDeou** [comment]
Related to #3199 

They could even be implemented with the same action.
Example:
```
# Pipe the scrollback buffer in foo.sh
keybind = ctrl+t=command:pipe:scrollback:foo.sh
# Run bar.sh
keybind = crtr+shift+t=command:bar.sh
# Paste the output of bar.sh
keybind = crtr+alt+t=command:pipe:paste:bar.sh
```

**Comment by joshuafcole** [comment]
Relatedly, the suggestion of `ghostty -a <action>` or similar to programmatically invoke an action could help early ghostty users (or those running niche distributions) in the transitional period where it slowly develops native integrations. For example, since most distros offer some method of registering global binds at the WM/DM level, linux users could use that in lieu of `global:` bindings support.

**Comment by jechter** [comment]
I would also be interested in a method to programmatically invoke ghostty keybind actions - specifically write_scrollback_file. I wrote a tool to supply command line prompt suggestions (https://github.com/jechter/boltprompt), which has an option for LLM-based suggestions using the current terminal scrollback for context. I can get scrollback contents in Terminal.app and iTerm using AppleScript, but have not found a way to programmatically access it in ghostty yet.

**Comment by rhodes-b** [comment]
Ghostty plans on having the ability to be programmatically controlled in the future the discussion that has the most information about this (since implementing it requires this ability) is #2353

You may also be interested in #3782 but that just leads back to the other issue I linked above

TL;DR this is mostly a dupe of what #2353 wants to implement (it goes beyond just windows, tabs and splits).

I didnt read exactly what the proposals here are but its very nice to consolidate into one discussion please use #2353 going forward

**Comment by jokeyrhyme** [comment]
I've turned off window decorations because GTK and Adwaita are subjectively not a fit for my setup: I'm inconsistently fussy :)

Anyhow, this means I lose the built-in tabs feature, so I've had to run `zellij` ( https://zellij.dev/ ) to bring them back (this probably makes Ghostty's creator very sad, I'm sorry)

And then to make the Ctrl+Tab and Ctrl+Shift+Tab keys switch `zellij` tabs, I can configure these bindings in `alacritty` to run the custom `zellij ...` commands to control the tabs, and this works

But there's not currently a way to map Ghostty bindings to run custom commands, so I've lost the ability to switch tabs when I'm running `zellij` in Ghostty

**Comment by jokeyrhyme** [comment]
Maybe related to https://github.com/ghostty-org/ghostty/discussions/4406 ?

**Comment by stevenwalton** [comment]
Just want to add my vote here. In Wezterm I have a [keybinding](https://github.com/stevenwalton/.dotfiles/blob/master/configs/wezterm/commands.lua#L68-L72) to toggle [Caffeine](https://www.caffeine-app.net/) by calling an [osascript](https://github.com/stevenwalton/.dotfiles/blob/master/scripts/OSX/caffeine.scpt).

Having this feature has a lot of benefits, especially if we can continue to use the `global` bindings. Seems to have a huge potential.

If these scripts need to run in a terminal, then maybe they could be run in the quickterm (thus if they error they error there) or maybe in a [pop-up](https://github.com/ghostty-org/ghostty/discussions/3197) (with similar error message handling).

Related, it may be good to have OS dependent options ([how I did this in Wez](https://github.com/stevenwalton/.dotfiles/blob/master/configs/wezterm/helpers.lua#L9-L27). Refer to the `keybindings` link)

**Comment by rhodes-b** [comment]
# BEFORE POSTING HERE PLEASE READ THIS

Like stated in my previous comment the feature request here will be duplicate work of the one in #2353 Just be cause the title says windows, tabs and splits it goes beyond that as the system that allows that would allow allow scripting from shell like setting keybinds **please stop** posting on this issue and use #2353 so everything can be consolidated in one issue

**Reply by stevenwalton** [reply]
Forgive me, but I did not make that connection, even after reading through the comments. I see a lot of talk about scripting ghostty, but not calling scripts. These are very different things. I can see if the solution is to call new surfaces, run the script there, and then close them, but if that's the plan then I think it would help to say that explicitly so others are not confused. I'm just taking a guess here due to your comment. Because even with your comment, #2353 looks like a completely different topic. I think if this was clarified you could help reduce dupes.

**Reply by rhodes-b** [reply]
#2353 references kitty remote control and iterm2 scripting, which allows you to trigger commands,  your keybinds in the config would use this remote protocol to ask ghostty to do stuff for you for example

**Reply by stevenwalton** [reply]
Just to be clear (so we can point people in the right direction), is the main comment to follow [this one](https://github.com/ghostty-org/ghostty/discussions/2353#discussioncomment-11676997)? There's a lot going on in that discussion and not much threading so I don't expect people searching for issues to connect them unless they read the entire thing with the right context (no one is going to read the entire thread when searching for issues).

**Reply by jokeyrhyme** [reply]
I'm confused, I thought this was about having ghostty keybinds trigger arbitrary executables/scripts

I'm not interested in controlling ghostty from outside of ghostty otherwise

**Reply by rhodes-b** [reply]
> ghostty -a write_scrollback_file:paste

Is a given example. `ghostty` is still a separate binary that you run. That means it is a separate process. This will then require that it can somehow communicate with this parent ghostty process on how to get the scrollback file which is what #2353 is trying to solve

**Comment by 00-kat** [comment]
> It would be awesome if ghostty supported more complex keybind actions

Is this discussion about complex keybind actions in general?  Or is it about a way to launch an arbitrary program via a keybinding?  If the latter, it would be nice if the discussion title was renamed to be more specific, since it's pretty confusing what this is about.  And if the former, I'm not sure if having a general catch-all discussion for "basically any keybind action people want" is a good idea.

(Of course, if it's to launch an arbitrary program via a keybinding, the best possible resolution is for every DE/WM/OS to have a way to scope keybindings to specific applications instead of every application implementing something like this, but everyone knows how likely that is to happen.)

**Reply by pluiedev** [reply]
I'm highly skeptical also of expanding keybind functions to run arbitrary commands. Ghostty is not a script runner.

**Reply by 00-kat** [reply]
Albeit the usefulness of this is limited without #2353, an ability to run an arbitrary command would allow for very interesting keybindings which can in the end control Ghostty.  Though without #2353 this ability would just be an elaborate way to have application-scoped keybindings due to inabilities of operating systems/desktop environments.

**Comment by mhinsch** [comment]
As far as I understood it the idea is to open the ghostty API to a shell
interface (as kitty has it for example), so that shell scripts could for
example open a new split.

**Reply by pluiedev** [reply]
That is #2353.
